File: ported/Voronoi.py
__author__ = 'sean'

from ported.VParabola import VParabola
from VQueue import VQueue
from VEvent import VEvent
from VPolygon import VPolygon
from VEdge import VEdge
from Point import Point, distance
import sys
class Voronoi:
    def __init__(self):
        self.places = None
        self.edges = None
        self.cells = None
        self.queue = VQueue()

        self.width = 0
        self.height = 0
        self.root = None
        self.ly = 0
        self.lasty = 0
        self.fp = None
    def Compute(self, p, width, height):
        if len(p) < 2:
            return []
        self.root = None
        self.places = p
        self.edges = []
        self.cells = []
        self.width = width
        self.height = height
        self.queue.clear()

        for p in self.places:
            ev = VEvent(p, True)
            cell = VPolygon()
            p.cell = cell
            self.queue.enqueue(ev)
            self.cells.append(cell)

        lasty = sys.float_info.max
        num = 0
        while not self.queue.isEmpty():
            e = self.queue.dequeue()
            self.ly = e.point.y
            if e.pe:
                self.InsertParabola(e.point)
            else:
                selfl.RemoveParabola(e)
            self.lasty = e.y
        self.FinishEdge(self.root)
        for e in self.edges:
            if e.neighbour:
                e.start = e.neighbour.end
    def GetEdges(self):
        return self.edges
    def GetCells(self):
        return self.cells

    #METHODS TO WORK WITH THE TREE
    
    def InsertParabola(self, p):
        if not self.root:
            self.root = VParabola(p)
            self.fp = p
            return None
        if self.root.isLeaf and self.root.site.y - p.y < 0.01:
            #degenerated parabola
            self.root.isLeaf = False
            self.root.left = VParabola(self.fp)
            self.root.right = VParabola(p)
            s = Point((p.x + self.fp.x) / 2.0, self.height)
            if p.x > self.fp.x:
                self.root.edge = VEdge(s, self.fp, p)
            self.edges.append(self.root.edge)
            return None

        par = self.getParabolaByX(p.x)

        if par.cEvent:
            self.queue.remove(par.cEvent)
            par.cEvent = None

        start = Point(p.x, self.GetY(par.site, p.x))

        el = VEdge(start, par.site, p)
        er = VEdge(start, p, par.site)
        #TODO:  REST OF FUNCTION
    def RemoveParabola(self, e):
        p1 = e.arch

        xl = self.GetLeftParent(p1)
        xr = self.GetRightParent(p1)

        p0 = self.GetLeftChild(xl)
        p2 = self.GetRightChild(xr)

        if p0.cEvent:
            self.queue.remove(p0.cEvent)
            p0.cEvent = None
        if p2.cEvent:
            self.queue.remove(p0.cEvent)
            p2.cEvent = None

        p = Point(e.point.x, self.GetY(p1.site, e.point.x))
        #TODO:  comparison of points is not implemented?
        if p0.site.cell.last == p1.site.cell.first:
            p1.site.cell.addLeft(p)
        else:
            p1.site.cell.addRight(p)

        p0.site.cell.addRight(p)
        p2.site.cell.addLeft(p)

        self.lasty = e.point.y

        xl.edge.end = p
        xr.edge.end = p

        higher = None
        par = p1
        while par != self.root:
            par = par.parent
            if par == xl:
                higher = xl
            if par == xr:
                higher = xr

        higher.edge = VEdge(p, p0.site, p2.site)

        self.edges.append(higher.edge)

        gparent = p1.parent.parent
        if p1.parent.left == p1:
            if gparent.left == p1.parent:
                gparent.left = p1.parent.right
            else:
                p1.parent.parent.right = p1.parent.right
        else:
            if gparent.left == p1.parent:
                gparent.left = p1.parent.left
            else:
                gparent.right = p1.parent.left
        self.CheckCircle(p0)
        self.CheckCircle(p2)
    def FinishEdge(self, n):
        mx = None
        if n.edge.direction.x > 0.0:
            mx = max((self.width, n.edge.start.x + 10))
        else:
            mx = min((0.0, n.edge.start.x - 10))
        n.edge.end = Point(mx, n.edge.f * mx + n.edge.g)
        if not n.left.isLeaf:
            self.FinishEdge(n.left)
        if not n.right.isLeaf:
            self.FinishEdge(n.right)
    def getXOfEdge(self, par, y):
        left = self.getLeftChild(par)
        right = self.GetRightChild(par)
        p = left.site
        r = right.site

        dp = 2.0 * (p.y - y)
        a1 = 1.0 / dp
        b1 = -2.0 * p.x / dp
        c1 = y + dp * 0.25 + p.x * p.x / dp

        dp = 2.0 * (r.y - y)
        a2 = 1.0 / dp
        b2 = -2.0 * r.x / dp
        c2 = y + dp * 0.25 + r.x * r.x / dp

        a = a1 - a2
        b = b1 - b2
        c = c1 - c2

        disc = b * b - 4 * a * c
        x1 = (-b + math.sqrt(disc)) / (2.0 * a)
        x2 = (-b - math.sqrt(disc)) / (2.0 * a)

        ry = None
        if p.y < r.y:
            ry = max((x1,x2))
        else:
            ry = min((x1,x2))
        return ry

    def getParabolaByX(self, xx):
        par = self.root
        x = 0
        while not par.isLeaf:
            x = self.getXOfEdge(par, self.ly)
            if x > xx:
                par = par.left
            else:
                par = par.right
        return par
    def GetY(self, p, x):
        dp = 2.0 * (p.y - self.ly)
        b1 = -2.0 * p.x / dp
        c1 = self.ly + dp / 4.0 + p.x * p.x / dp
        return x*x / dp + b1 * x + c1
    def CheckCircle(self, b):
        lp = self.GetLeftParent(b)
        rp = self.GetRightParent(b)
        a = self.GetLeftChild(lp)
        c = self.GetRightChild(rp)

        if not a or not c or a.site == c.site:
            return None
        d = distance(a.site, s)
        if s.y - d >= self.ly:
            return
        e = Event(Point(s.x, s.y - d), False)
        b.cEvent = e
        e.arch = b
        self.queue.enqueue(e)
    def GetEdgeIntersection(self, a, b):
        I = self.GetLineIntersection(a.star, a.B, b.start, b.B)
        #wrong direction of edge
        wd = (I.x - a.start.x) * a.direction.x < 0 \
            or (I.y - a.start.y) * a.direction.y < 0 \
                or (I.x - b.start.x) * b.direction.x < 0 \
                    or (I.y - b.start.y) * b.direction.y < 0
        if wd:
            return None
        return I
    def GetLeft(self, n):
        return self.GetLeftChild(self.GetLeftParent(n))
    def GetRight(self, n):
        return self.GetRightChild(self.GetRightParent(n))
    def GetLeftParent(self, n):
        par = n.parent
        pLast = n
        while par.left == pLast:
            if not par.parent:
                return None
            pLast = par
            par = par.parent
        return par
    def GetRightParent(self, n):
        par = n.parent
        pLast = n
        #TODO:  need to write comparison function
        while par.right == pLast:
            if not par.parent:
                return None
            pLast = par
            par = par.parent
        return par
    def GetLeftChild(self, n):
        if not n:
            return None
        par = n.left
        while not par.isLeaf:
            par = par.right
        return par
    def GetRightChild(self, n):
        if not n:
            return None
        par = n.right
        while not par.isLeaf:
            par = par.left
        return par
    def GetLineIntersection(self, a1, a2, b1, b2):
        dax = (a1.x - a2.x)
        dbx = (b1.x - b2.x)
        day = a1.y - a2.y
        dby = b1.y - b2.y
        Den = dax*dby - day * dbx
        if Den == 0:
            return None
        A = (a1.x * a2.y - a1.y - a2.x)
        B = b1.x * b2.y - b1.y * b2.x
        I = Point(0,0)
        I.x = (A * dbx - B * dax) / Den
        I.y = (A * dby - B * day) / Den
        return I

File: javascript/Voronoi.js
function Voronoi()
{
	with(this)
	{
		this.places = null;
		this.edges = null;
		this.cells = null;
		this.queue = new VQueue;
		
		this.width = 0;
		this.heght = 0;
		this.root = null;
		this.ly = 0;
		this.lasty = 0;
		this.fp = null;
	}
}

Voronoi.prototype.Compute = function(p, width, height)
{
	if(p.length<2) return [];

	this.root = null;
	this.places = p;
	this.edges = [];
	this.cells = [];
	this.width = width;
	this.height = height;
	
	this.queue.clear(true);
	
	for(i=0; i<this.places.length; i++)
	{
		var ev = new VEvent(this.places[i], true);
		var cell = new VPolygon();
		this.places[i].cell = cell;
		this.queue.enqueue(ev);
		this.cells.push(cell);
	}
	
	var lasty = Number.MAX_VALUE;
	var num = 0;
	while(!this.queue.isEmpty())
	{
		var e = this.queue.dequeue();  
		this.ly = e.point.y;
		if(e.pe) this.InsertParabola(e.point);
		else this.RemoveParabola(e);
		
		this.lasty = e.y;
	}
	this.FinishEdge(this.root);
	
	for(i=0; i<this.edges.length; i++)
		if(this.edges[i].neighbour) this.edges[i].start = this.edges[i].neighbour.end;
}

Voronoi.prototype.GetEdges = function()
{
	return this.edges;
}

Voronoi.prototype.GetCells = function()
{
	return this.cells;
}

// M E T H O D S   F O R   W O R K   W I T H   T R E E -------
		
Voronoi.prototype.InsertParabola = function(p)
{
	if(!this.root){this.root = new VParabola(p); this.fp = p; return;}
	
	if(this.root.isLeaf && this.root.site.y - p.y <0.01)	// degenerovaný případ - první dvě místa ve stejné výšce
	{
		this.root.isLeaf = false;
		this.root.left = new VParabola(this.fp);
		this.root.right = new VParabola(p);
		var s = new Point((p.x+this.fp.x)/2, this.height);
		if(p.x>this.fp.x) this.root.edge = new VEdge(s, this.fp, p);
		else this.root.edge = new VEdge(s, p, this.fp);
		this.edges.push(this.root.edge);
		return;
	}
	
	var par = this.GetParabolaByX(p.x);
	
	if(par.cEvent)
	{
		this.queue.remove(par.cEvent);
		par.cEvent = null;
	}

	var start = new Point(p.x, this.GetY(par.site, p.x));
	
	var el = new VEdge(start, par.site, p);
	var er = new VEdge(start, p, par.site);
	
	el.neighbour = er;
	this.edges.push(el);
	
	par.edge = er;
	par.isLeaf = false;
	
	var p0 = new VParabola(par.site);
	var p1 = new VParabola(p);
	var p2 = new VParabola(par.site);
	
	par.right = p2;
	par.left = new VParabola();
	par.left.edge = el;

	par.left.left = p0;
	par.left.right = p1;
	
	this.CheckCircle(p0);
	this.CheckCircle(p2);
}
		
Voronoi.prototype.RemoveParabola = function(e)
{						
	var p1 = e.arch;
	
	var xl = this.GetLeftParent(p1);
	var xr = this.GetRightParent(p1);
		
	var p0 = this.GetLeftChild(xl);
	var p2 = this.GetRightChild(xr);
	
	if(p0.cEvent){this.queue.remove(p0.cEvent); p0.cEvent = null;}
	if(p2.cEvent){this.queue.remove(p2.cEvent); p2.cEvent = null;}
				
	var p = new Point(e.point.x, this.GetY(p1.site, e.point.x));

	
	if(p0.site.cell.last == p1.site.cell.first ) p1.site.cell.addLeft(p);
	else p1.site.cell.addRight(p);
	
	p0.site.cell.addRight(p);
	p2.site.cell.addLeft(p);
	
	this.lasty = e.point.y;
		
	xl.edge.end = p;
	xr.edge.end = p;
	
	var higher;
	var par = p1;
	while(par != this.root)
	{
		par = par.parent;
		if(par == xl) {higher = xl;}
		if(par == xr) {higher = xr;}
	}
	
	higher.edge = new VEdge(p, p0.site, p2.site);

	this.edges.push(higher.edge);
	
	var gparent = p1.parent.parent;
	if(p1.parent.left == p1)
	{
		if(gparent.left  == p1.parent) gparent.left  = p1.parent.right;
		else p1.parent.parent.right = p1.parent.right;
	}
	else
	{
		if(gparent.left  == p1.parent) gparent.left  = p1.parent.left;
		else gparent.right = p1.parent.left;
	}
	
	this.CheckCircle(p0);
	this.CheckCircle(p2)
}

Voronoi.prototype.FinishEdge = function(n)
{
	var mx;
	if(n.edge.direction.x > 0.0)
	{
		mx = Math.max(this.width, n.edge.start.x + 10 );
	}
	else
	{
		mx = Math.min(0.0, n.edge.start.x - 10);
	}
	n.edge.end = new Point(mx, n.edge.f*mx + n.edge.g);
	
	if(!n.left.isLeaf)  this.FinishEdge(n.left);
	if(!n.right.isLeaf) this.FinishEdge(n.right);
}

Voronoi.prototype.GetXOfEdge = function(par, y) // počítá průsečík parabol v daném uzlu
{
	var left =	this.GetLeftChild (par);
	var right =	this.GetRightChild(par);
			
	var p = left.site;
	var r = right.site;
	
	var dp = 2*(p.y - y);
	var a1 = 1/dp;
	var b1 = -2*p.x/dp;
	var c1 = y+dp*0.25 + p.x*p.x/dp;
	
	dp = 2*(r.y - y);
	var a2 = 1/dp;
	var b2 = -2*r.x/dp;
	var c2 = y+dp*0.25 + r.x*r.x/dp;
	
	var a = a1 - a2;
	var b = b1 - b2;
	var c = c1 - c2;
	
	var disc = b*b - 4 * a * c;
	var x1 = (-b + Math.sqrt(disc)) / (2*a);
	var x2 = (-b - Math.sqrt(disc)) / (2*a);

	var ry;
	if(p.y < r.y ) ry =  Math.max(x1, x2);
	else ry = Math.min(x1, x2);
	
	return ry;
}

Voronoi.prototype.GetParabolaByX = function(xx)
{
	var par = this.root;
	var x = 0;
	
	while(!par.isLeaf)
	{
		x = this.GetXOfEdge(par, this.ly);
		if(x>xx) par = par.left;
		else par = par.right;				
	}
	return par;
}

Voronoi.prototype.GetY = function(p, x) // ohnisko, x-souřadnice, řídící přímka
{
	var dp = 2*(p.y - this.ly);
	var b1 = -2*p.x/dp;
	var c1 = this.ly+dp/4 + p.x*p.x/dp;
	
	return(x*x/dp + b1*x + c1);
}

Voronoi.prototype.CheckCircle = function(b)
{
	var lp = this.GetLeftParent(b);
	var rp = this.GetRightParent(b);
	
	var a = this.GetLeftChild(lp);
	var c = this.GetRightChild(rp);
	
	if(!a || !c || a.site == c.site) return;
	
	var s = this.GetEdgeIntersection(lp.edge, rp.edge);
	if(!s) return;
	
	var d = Point.prototype.distance(a.site, s);
	//if(d > 5000) return;
	if(s.y - d  >= this.ly) return;
	
	var e = new VEvent(new Point(s.x, s.y - d), false);
	
	b.cEvent = e;
	e.arch = b;
	this.queue.enqueue(e);
}

Voronoi.prototype.GetEdgeIntersection = function(a, b)
{
	var I = GetLineIntersection(a.start, a.B, b.start, b.B);
	
	// wrong direction of edge
	var wd = 	(I.x - a.start.x)*a.direction.x<0 || (I.y - a.start.y)*a.direction.y<0	
			 ||	(I.x - b.start.x)*b.direction.x<0 || (I.y - b.start.y)*b.direction.y<0;	
			 
	if(wd) return null;
	return I;
}

Voronoi.prototype.GetLeft = function(n)
{
	return this.GetLeftChild( this.GetLeftParent(n));
}
		
Voronoi.prototype.GetRight = function(n)
{
	return this.GetRightChild(this.GetRightParent(n));
}	
		
Voronoi.prototype.GetLeftParent = function(n)
{
	var par = n.parent;
	var pLast = n;
	while(par.left == pLast) 
	{ 
		if(!par.parent) return null;
		pLast = par; par = par.parent; 
	}
	return par;
}

Voronoi.prototype.GetRightParent = function(n)
{
	var par = n.parent;
	var pLast = n;
	while(par.right == pLast) 
	{	
		if(!par.parent) return null;
		pLast = par; par = par.parent;	
	}
	return par;
}
Voronoi.prototype.GetLeftChild = function(n)
{
	if(!n) return null;
	var par = n.left;
	while(!par.isLeaf) par = par.right;
	return par;
}
Voronoi.prototype.GetRightChild = function(n)
{
	if(!n) return null;
	var par = n.right;
	while(!par.isLeaf) par = par.left;
	return par;
}

function GetLineIntersection(a1, a2, b1, b2)
{			
	var dax = (a1.x-a2.x), dbx = (b1.x-b2.x);
	var day = (a1.y-a2.y), dby = (b1.y-b2.y);
			
	var Den = dax*dby - day*dbx;
	if (Den == 0) return null;	// parallel

	var A = (a1.x * a2.y - a1.y * a2.x);
	var B = (b1.x * b2.y - b1.y * b2.x);
		
	var I = new Point(0,0);
	I.x = ( A*dbx - dax*B ) / Den;
	I.y = ( A*dby - day*B ) / Den;
	
	return I;
}




